### ISSUE TITLE
BUG: scipy 1.9 build issue

### ISSUE THREAD

**Issue by xkszltl** [scipy/scipy#17185]
### Describe your issue.

We build scipy and scikit-learn from source.
With scipy 1.9, scikit-learn build always failed immediately with log below.
We also tried `python3 -c 'from scipy.lib import _ccallback_c'` and that failed in the same way.

This wasn't an issue in 1.8 or even earlier version.

If we build lower version first and gradually upgrade, it works fine, indicating probably some file is left there.

### Reproducing Code Example

```python
Bascially python3 -m pip install on locally cloned repo of scipy/scikit-learn
```

### Error message

```shell
#12 247.1   Traceback (most recent call last):                                                                                                                                                        
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/scipy/__init__.py", line 166, in <module>                                                                                                  
#12 247.1       from scipy._lib._ccallback import LowLevelCallable                                                                                                                                    
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/scipy/_lib/_ccallback.py", line 1, in <module>                                                                                             
#12 247.1       from . import _ccallback_c                                                                                                                                                            
#12 247.1   ImportError: cannot import name '_ccallback_c' from 'scipy._lib' (/usr/local/lib/python3.9/dist-packages/scipy/_lib/__init__.py)                                                          
#12 247.1                                                                                                                                                                                             
#12 247.1   The above exception was the direct cause of the following exception:                                                                                                                      
#12 247.1                                                                                                                                                                                             
#12 247.1   Traceback (most recent call last):                                                                                                                                                        
#12 247.1     File "setup.py", line 200, in check_package_status                                                                                                                                      
#12 247.1       module = importlib.import_module(package)                                                                                                                                             
#12 247.1     File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module                                                                                                             
#12 247.1       return _bootstrap._gcd_import(name[level:], package, level)                                                                                                                           
#12 247.1     File "<frozen importlib._bootstrap>", line 1030, in _gcd_import                                                                                                                         
#12 247.1     File "<frozen importlib._bootstrap>", line 1007, in _find_and_load                                                                                                                      
#12 247.1     File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked                                                                                                              
#12 247.1     File "<frozen importlib._bootstrap>", line 680, in _load_unlocked                                                                                                                       
#12 247.1     File "<frozen importlib._bootstrap_external>", line 790, in exec_module                                                                                                                 
#12 247.1     File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed                                                                                                            
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/scipy/__init__.py", line 171, in <module>                                                                                                  
#12 247.1       raise ImportError(msg) from e                                                                                                                                                         
#12 247.1   ImportError: The `scipy` install you are using seems to be broken, (extension modules cannot be imported), please try reinstalling.                                                       
#12 247.1   Traceback (most recent call last):                                                                                                                                                        
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>                                                                       
#12 247.1       main()                                                                                                                                                                                
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main                                                                           
#12 247.1       json_out['return_val'] = hook(**hook_input['kwargs'])                                                                                                                                 
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/pip/_vendor/pep517/in_process/_in_process.py", line 164, in prepare_metadata_for_build_wheel                                               
#12 247.1       return hook(metadata_directory, config_settings)                                                                                                                                      
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/setuptools/build_meta.py", line 188, in prepare_metadata_for_build_wheel                                                                   
#12 247.1       self.run_setup()                                                                                                                                                                      
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/setuptools/build_meta.py", line 281, in run_setup                                                                                          
#12 247.1       super(_BuildMetaLegacyBackend,                                                                                                                                                        
#12 247.1     File "/usr/local/lib/python3.9/dist-packages/setuptools/build_meta.py", line 174, in run_setup                                                                                          
#12 247.1       exec(compile(code, __file__, 'exec'), locals())                                                                                                                                       
#12 247.1     File "setup.py", line 329, in <module>                                                                                                                                                  
#12 247.1       setup_package()                                                                                                                                                                       
#12 247.1     File "setup.py", line 305, in setup_package                                                                                                                                             
#12 247.1       check_package_status("scipy", min_deps.SCIPY_MIN_VERSION)                                                                                                                             
#12 247.1     File "setup.py", line 227, in check_package_status                                                                                                                                      
#12 247.1       raise ImportError(                                                                                                                                                                    
#12 247.1   ImportError: scipy is not installed.
```

### SciPy/NumPy/Python version information

Latest/Latest/3.8

**Comment by xkszltl** [comment]
1.9.0 may also work, but not 1.9.1 and 1.9.2

**Comment by tupui** [comment]
Hi @xkszltl thank you for reporting. Can you give more details about your setup and how you are building SciPy? Are you following our development instructions?

**Comment by rgommers** [comment]
Can you add `-vv` to your `pip install` command so that we can see the actual issue in the full build log? 

> `from . import _ccallback_c`

That is what is failing. Does `scipy/_lib/_ccallback_c.cpython-<xxx>.so` actually exist? If so, can you import it directly (e.g., `cd scipy/_lib`; `python -c "import _ccallback_c"`)?

**Comment by xkszltl** [comment]
> We also tried `python3 -c 'from scipy.lib import _ccallback_c'` and that failed in the same way.

As mentioned the import is failing regardless of manually done or as part of scikit-learn build.
The .so file is not there.

For `-vv` I can do that but where should I look for the error?
scipy build itself completed successfully but without generating that file, the one failing is scikit-learn build and the error msg is simply the import failure above.

**Comment by xkszltl** [comment]
> Hi @xkszltl thank you for reporting. Can you give more details about your setup and how you are building SciPy? Are you following our development instructions?

These lines:
- https://github.com/xkszltl/Roaster/blob/387bd9c052b6964042affbefd80773444fc61514/pkgs/utils/fpm/distro_cc.sh#L16
- `export GIT_REPO=https://github.com/scipy/scipy.git GIT_TAG=v1.9.2`
- https://github.com/xkszltl/Roaster/blob/387bd9c052b6964042affbefd80773444fc61514/pkgs/utils/pip_install_from_git.sh#L116
- https://github.com/xkszltl/Roaster/blob/387bd9c052b6964042affbefd80773444fc61514/pkgs/utils/pip_install_from_git.sh#L132

Basically `sudo -E CC=gcc-10 CXX=g++-10 FC=gfortran-10 python3 -m pip install --no-clean -Uv ./scipy` with a freshly re-cloned dir every time.

**Comment by rgommers** [comment]
> For `-vv` I can do that but where should I look for the error?
> scipy build itself completed successfully but without generating that file,

The extension module must have been built, but then failed to install somehow. There's lots of lines like this:
```
  [1/2] Generating scipy/generate-version with a custom command
  fatal: bad revision '^v1.8.0'
  [2/2] Generating scipy/generate-config with a custom command
  Installing scipy/__config__.py to /home/runner/work/scipy/scipy/.mesonpy-2_3ry3af/install/usr/lib/python3/dist-packages/scipy
  Installing scipy/version.py to /home/runner/work/scipy/scipy/.mesonpy-2_3ry3af/install/usr/lib/python3/dist-packages/scipy
  Installing scipy/_lib/_ccallback_c.cpython-310-x86_64-linux-gnu.so to /home/runner/work/scipy/scipy/.mesonpy-2_3ry3af/install/usr/lib/python3/dist-packages/scipy/_lib
...
<one line per installed file, ~1200 files>
```
There will be warnings visible there (capitalized, `WARNING ...`) if something is not installed.

Posting the full install log would be helpful.



> Basically `sudo -E CC=gcc-10 CXX=g++-10 FC=gfortran-10 python3 -m pip install --no-clean -Uv ./scipy` with a freshly re-cloned dir every time.

Ah, `sudo` is an untested thing here. Best practice is to never do that. Perhaps something is going wrong due to that.

**Comment by xkszltl** [comment]
I know python community generally suggest to use venv or whatever isolated env and discourage the use of sudo, but in our cases we often need to interact with native libs or talk with other non-python component in system, so we prefer to use the distro-stock python and sudo install to `/usr{,/local}/lib{,64}/python3.*/{dist,site}-packages`.

**Comment by rgommers** [comment]
Yes, I know there's reasons you may have to use `sudo`. If that's the cause, I do want to fix it. Just saying that it's untested so far, so may be the root cause somehow.

**Comment by xkszltl** [comment]
Emmm, even though it was stably reproducible the day I filed this issue, across multiple machines and dockers, it's gone the next day and I cannot repro it at all.
So at this point I'm very confused and have no idea what's going on.

The only "timely" thing I recall that day is numpy tagged 1.23.4 but haven't push wheel to pypa, we built numpy from source and use the latest release tag, and some `pip install <git_repo>` calls complained about numpy 1.23.4 doesn't exists.

Another possibly related thing is, we manually cleared ccache used by some docker build, although that cannot explained what happened elsewhere.

**Comment by rgommers** [comment]
> Another possibly related thing is, we manually cleared ccache used by some docker build, although that cannot explained what happened elsewhere.

That could indeed explain an issue going away. You filed the issue on Oct 9th, while there were bug fix releases for both `meson` and `meson-python` on Oct 5th with fixes relevant to SciPy builds.

Do you want to test more, or close this issue now and reopen in case it comes back?

**Comment by xkszltl** [comment]
I can close now and reopen if I got this again, thanks again for the quick help~!
